### DISCUSSION TITLE
Pretrain Encoder Network using Reconstruction VAE or AE

### DISCUSSION THREAD

**Discussion by nicozhou7** [open-edge-platform/anomalib#786] [Feature Requests] [answered]
Hi, I want to request a feature to pretrain an encoder network to adapt to the dataset before using the models like patchcore and fastflow. Is there any workaround for that?

**Comment by ashwinvaidya17** [comment, marked as answer]
@nixczhou We currently have no plans to add pre-training to the Anomalib pipeline. However, you can always pre-train the model externally and load the weights into Anomalib. There are a couple of ways to achieve this.

1. You can manually load the weights by using `torch.load` in the lightning model or the model checkpoint callback. https://github.com/openvinotoolkit/anomalib/blob/942cada16b42f8cbc2b88b34fd34aa4b3c4cf9ef/anomalib/utils/callbacks/model_loader.py#L19 Model checkpoint callback expects the entire lightning module as a single weight file but you can always work around that.
2. We are also modifying our feature extractor component (https://github.com/openvinotoolkit/anomalib/pull/748/). When this is merged you can then do something like this.
   ```python
   class CustomTorchModel(nn.Module):
        def __init__(self, ...):
            self.encoder = get_feature_extractor(backbone=BackboneClass, return_nodes=["embedding_layer"], weights="pretrained_weights.pth")
            ...
    ...
   ```

**Reply by nicozhou7** [reply]
Does the LoadModelCallback also work for changing the Backbone  for the pretrain model used in for example Patchcore model?

**Reply by ashwinvaidya17** [reply]
`LoadModelCallback` currently loads the weights for the entire model. This means the backbone and any other layers of the model. But you can  always implement some custom logic in the `LoadModelCallback` class to load the backbone from a different weight file.
As for changing the backbone of the feature extractor, the current limitation is that you can only use pre-trained models from Timm. But you can change this line in Patchcore and instead of calling `FeatureExtractor`, you can implement your own method to load a custom backbone.
https://github.com/openvinotoolkit/anomalib/blob/32c2d195fba514b307a3c47a864c965fb20f0c00/anomalib/models/patchcore/torch_model.py#L40
You can also look into replacing this line with [TorchFXFeatureExtractor](https://github.com/openvinotoolkit/anomalib/blob/32c2d195fba514b307a3c47a864c965fb20f0c00/anomalib/models/components/feature_extractors/torchfx.py) as it gives you more flexibility. You can pass custom model class and weights to it.

Hope this is clear.
